### PR TITLE
Lusine: Tasks 171 _180

### PR DESCRIPTION
**Pull request by hsp-lusine**

### CODE REVIEW

File: src/main/java/am/hsp/apps/seua/task171_180/Task_177.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * Count and Print the following sequence, where is x(1) equal NUMBER_1 and x(2)
+ * equal NUMBER_2;
+ * 
+ * @author lusine
+ *
+ */
+public class Task_177 extends TaskDigitBase {
+	private static double NUMBER_1 = 1;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_177.java]: same

File: src/main/java/am/hsp/apps/seua/task171_180/Task_171.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	 * Calculates and returns a factorial with method Recursion the given number
+	 * 
+	 * @param number
+	 * @return
+	 */
+	private int getFactoialWithRecursion(int number) {
+		if (number == 1 || number == 0) {
+			return 1;
+		}
+		return (number >= 1) ? getFactoialWithRecursion(number - 1) * number : NOT_FOUND;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_171.java]: Wrong

File: src/main/java/am/hsp/apps/seua/task171_180/Task_174.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	 * 
+	 * @param number
+	 */
+	private double getResult(int number) {
+        double current = FIRST_NUMBER;
+		double previous = FIRST_NUMBER;
+
+		for (int i = 1; i < number; i++) {
+		   current = (2 + 1 / previous);
+			previous = current;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_174.java]: So where are you printing "**current**" ???? You MUST print **current** element here!

File: src/main/java/am/hsp/apps/seua/task171_180/Task_174.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+           return current;
+	}
+
+	/**
+	 * Count and Return the following sequence with Recursion: x[n] = 2 + 1 / x[n-1]
+	 * 
+	 * @param index
+	 * @return
+	 */
+	private double getCurrentElement(int index) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_174.java]: It is better to update method name by indicating that here you are calculating current via recursion:
**getCurrentElementRecursion**

File: src/main/java/am/hsp/apps/seua/task171_180/Task_174.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+		System.out.println();
+		printResultWithRecursion(number);
+	}
+
+	/**
+	 * Count and Return the following sequence: x[n] = 2 + 1 / x[n-1]
+	 * 
+	 * @param number
+	 */
+	private double getResult(int number) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_174.java]: I guess this is actually **printResult** method!

File: src/main/java/am/hsp/apps/seua/task171_180/Task_174.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	
+	/**
+	 * System.out.println x[n] = 2 + 1 / x[n-1] with normal method
+	 * 
+	 * @param number
+	 */
+	private void printResult(int number) {
+		System.out.println("Result");
+		for (int i = 1; i <= number; i++) {
+			System.out.println(getResult(i));
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_174.java]: oh noooo...!!!!! Remove this method from here.
printResult actually getResult method in your current code, please see my comments on that method and make appropriate changes...

File: src/main/java/am/hsp/apps/seua/task171_180/Task_177.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	 * 
+	 * @see main.java.am.hsp.apps.seua.base.TaskBase#solve()
+	 */
+	@Override
+	public void solve() {
+		int number = getUserInput("Please enter number");
+
+		printSequence(number);
+		printSequenceWithRecursion(number);
+		;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_177.java]: ???

File: src/main/java/am/hsp/apps/seua/task171_180/Task_177.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
+package main.java.am.hsp.apps.seua.task171_180;
+
+import main.java.am.hsp.apps.seua.base.TaskBase;
+
+/**
+ * Count and Print the following sequence x[n] = (x[k-2] + 2*x[n-1]) / 3, where
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_177.java]: x[n - 2]

File: src/main/java/am/hsp/apps/seua/task171_180/Task_177.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+	public void solve() {
+		int number = getUserInput("Please enter number");
+
+		printSequence(number);
+		printSequenceWithRecursion(number);
+		;
+	}
+
+	/**
+	 * Print the following sequence: x[n] = (x[k-2] + 2*x[n-1]) / 3
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_177.java]: x[n - 2]

File: src/main/java/am/hsp/apps/seua/task171_180/Task_177.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+			double current = getCurrentElement(previous, prePrevious);
+			System.out.println(current);
+
+			prePrevious = previous;
+			previous = current;
+		}
+	}
+
+	/**
+	 * Calculates the current element with normal method
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_177.java]: That would be great if you will have here calculation formula

File: src/main/java/am/hsp/apps/seua/task171_180/Task_177.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+	 * 
+	 * @param previous
+	 * @return
+	 */
+	private double getCurrentElement(double number, double prevNumber) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_177.java]: number - bad param name

File: src/main/java/am/hsp/apps/seua/task171_180/Task_175.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+		for (int i = 1; i < number; i++) {
+			double current = getCurrentElement(previous, i);
+			System.out.println(current);
+			previous = current;
+		}
+		System.out.println();
+	}
+	
+	/**
+	 * Calculates the current element with normal method
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_175.java]: Would be great to have here current calculation formula

File: src/main/java/am/hsp/apps/seua/task171_180/Task_176.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	/**
+	 * Calculate the current Fibonacci Number
+	 * 
+	 * @param number
+	 * @param prevNumber
+	 * @return
+	 */
+	private int getCurrentFibonacciNumber(int number, int prevNumber) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_176.java]: number - bad param name

File: src/main/java/am/hsp/apps/seua/task171_180/Task_176.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+			System.out.print(current + ", ");
+
+			previous = prePrevious;
+			prePrevious = current;
+		}
+		System.out.println();
+	}
+
+	/**
+	 * Calculate the current Fibonacci Number
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_176.java]: same

File: src/main/java/am/hsp/apps/seua/task171_180/Task_178.java [hunk 15]
```diff
@@ -0,0 +1,106 @@
+package main.java.am.hsp.apps.seua.task171_180;
+
+import main.java.am.hsp.apps.seua.base.TaskBase;
+
+/**
+ * Count and Print the following sequence x[n] = x[k-1] + x[k-2] - 2*x[n-3],
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_178.java]: x[k -3]

File: src/main/java/am/hsp/apps/seua/task171_180/Task_178.java [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+		System.out.println();
+	}
+	
+	/**
+	 * Calculates the current element with normal method
+	 * 
+	 * @param previous
+	 * @return
+	 */
+	private int getCurrentElement(int prePreNumber, int prevNumber, int number) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_178.java]: wrong parameter names - all 3 param names are wrong

File: src/main/java/am/hsp/apps/seua/task171_180/Task_178.java [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+			
+			previous = prePrevious;
+			prePrevious = prePrePrevious;
+			prePrePrevious = current;
+		}
+		System.out.println();
+	}
+	
+	/**
+	 * Calculates the current element with normal method
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_178.java]: same

File: src/main/java/am/hsp/apps/seua/task171_180/Task_179.java [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+	/**
+	 * Use only "minus" and "plus", print the Number_N divison Number_k the received
+	 * number the Quotient and the Balance
+	 * 
+	 * @param number_n
+	 * @param number_k
+	 */
+	private void getQuotient(int number_n, int number_k) {
+		int quotient = 0;
+		while (number_n - number_k >= 0) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_179.java]: Please use do while, as currently you are using a "while" loop as "do/while".

File: src/main/java/am/hsp/apps/seua/task171_180/Task_179.java [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+	}
+
+	/**
+	 * Use only "minus" and "plus", print the Number_N divison Number_k the received
+	 * number the Quotient and the Balance
+	 * 
+	 * @param number_n
+	 * @param number_k
+	 */
+	private void getQuotient(int number_n, int number_k) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_179.java]: wrong name format for properties

File: src/main/java/am/hsp/apps/seua/task171_180/Task_177.java (hunk 11 above)
Comment (hsp-lusine) [reply src/main/java/am/hsp/apps/seua/task171_180/Task_177.java]: Can I write previousNumber and prePreviousNumber ??? previousNumber  but not previuos.... when call the method will be easier.

File: src/main/java/am/hsp/apps/seua/task171_180/Task_177.java [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+		}
+	}
+
+	/**
+	 * Calculates the current element with normal method (x[n-2] + 2*x[n-1]) / 3
+	 * 
+	 * @param previous
+	 * @return
+	 */
+	private double getCurrentElement(double previuosIndex, double prePreviuosIndex) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_177.java]: Again wrong param names. Please be attentive, my previous comment on this again was about wrong names.

File: src/main/java/am/hsp/apps/seua/base/TaskBase.java [hunk 21]
```diff
@@ -300,10 +302,50 @@ protected boolean isMultipleWithRemainder(int number, int divisor, int remainder
...
+
+	/**
+	 * Calculates and returns a factorial the given number
+	 * 
+	 * @param number
+	 * @return
+	 */
+	protected long getFactorial(int number) {
+		if (number < 0) {
+
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/base/TaskBase.java]: remove new line

File: src/main/java/am/hsp/apps/seua/base/TaskBase.java [hunk 22]
```diff
@@ -300,10 +302,50 @@ protected boolean isMultipleWithRemainder(int number, int divisor, int remainder
...
+			return NOT_FOUND;
+		}
+		int product = 1;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/base/TaskBase.java]: new line

File: src/main/java/am/hsp/apps/seua/base/TaskBase.java [hunk 23]
```diff
@@ -300,10 +302,50 @@ protected boolean isMultipleWithRemainder(int number, int divisor, int remainder
...
+	}
+
+	/**
+	 * Calculates and returns a double factorial the given number
+	 * 
+	 * @param number
+	 * @return
+	 */
+	protected long getDoubleFactorial(int number) {
+
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/base/TaskBase.java]: no new line required

File: src/main/java/am/hsp/apps/seua/base/TaskBase.java [hunk 24]
```diff
@@ -300,10 +302,50 @@ protected boolean isMultipleWithRemainder(int number, int divisor, int remainder
...
+		if (number < 0) {
+
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/base/TaskBase.java]: same

File: src/main/java/am/hsp/apps/seua/base/TaskBase.java [hunk 25]
```diff
@@ -300,10 +302,50 @@ protected boolean isMultipleWithRemainder(int number, int divisor, int remainder
...
+	 * @param number
+	 * @return
+	 */
+	protected long getDoubleFactorial(int number) {
+
+		if (number < 0) {
+			
+			return NOT_FOUND;
+		}
+		int product = 1;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/base/TaskBase.java]: new line from before and after. Please Read Oracle Code Conventions that I sent you!!

File: src/main/java/am/hsp/apps/seua/task171_180/Task_178.java [hunk 26]
```diff
@@ -0,0 +1,106 @@
+package main.java.am.hsp.apps.seua.task171_180;
+
+import main.java.am.hsp.apps.seua.base.TaskBase;
+
+/**
+ * Count and Print the following sequence x[n] = x[n-1] + x[k-2] - 2*x[n-3],
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_178.java]: You didn't fix my previous comment:
**x[k-3]**

File: src/main/java/am/hsp/apps/seua/task171_180/Task_176.java [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	/**
+	 * Calculate the current Fibonacci Number: F[n-1] + F[n-2]
+	 * 
+	 * @param previousIndex
+	 * @param prePreviousIndex
+	 * @return
+	 */
+	private int getCurrentFibonacciNumber(int previousIndex, int prePreviousIndex) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_176.java]: Wrong param names

File: src/main/java/am/hsp/apps/seua/task171_180/Task_178.java [hunk 28]
```diff
@@ -0,0 +1,106 @@
...
+		System.out.println();
+	}
+	
+	/**
+	 * Calculates the current element with normal method x[n-1] + x[n-2] - 2*x[n-3]
+	 * 
+	 * @param previous
+	 * @return
+	 */
+	private int getCurrentElement(int prePreIndex, int prevIndex, int index) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_178.java]: same.
your first argument is NOT previous of the previous element, but previous of previous of previous :)))
second param is previous of previous
third param is the previous

Please take a look on formula you written in the comment!!!!!

File: src/main/java/am/hsp/apps/seua/task171_180/Task_179.java [hunk 29]
```diff
@@ -0,0 +1,56 @@
...
+	 * Use only "minus" and "plus", print the Number_N divison Number_k the received
+	 * number the Quotient and the Balance
+	 * 
+	 * @param firstNumber
+	 * @param secondNumber
+	 */
+	private void getQuotient(int firstNumber, int secondNumber) {
+		int quotient = 0;
+		if (firstNumber < secondNumber) {
+			System.out.println("The Quotient = " + quotient + "\n" + "The Balance = " + firstNumber); return;
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_179.java]: wrong format

Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_179.java]: DRY violation

File: src/main/java/am/hsp/apps/seua/task171_180/Task_179.java [hunk 30]
```diff
@@ -0,0 +1,56 @@
...
+			
+		}
+		do {
+			firstNumber -= secondNumber;
+			quotient++;
+		} while (firstNumber - secondNumber >= 0);
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_179.java]: please update and improve your code

File: src/main/java/am/hsp/apps/seua/task171_180/Task_178.java [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+	}
+	
+	/**
+	 *Calculate current element with Recursion x[n-1] + x[n-2] - 2*x[n-3]
+	 * with Recursion
+	 * 
+	 * @param element
+	 * @return
+	 */
+	private int getCurrentElementWithRecursion(int element) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_178.java]: Wrong parameter name.

File: src/main/java/am/hsp/apps/seua/task171_180/Task_177.java [hunk 32]
```diff
@@ -0,0 +1,93 @@
...
+		}
+	}
+
+	/**
+	 * Calculate current element with Recursion (x[n-2] + 2*x[n-1]) / 3
+	 * 
+	 * @param element
+	 * @return
+	 */
+	private double getCurrentElementWithRecursion(int element) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_177.java]: Wrong parameter name.

File: src/main/java/am/hsp/apps/seua/task171_180/Task_176.java [hunk 33]
```diff
@@ -0,0 +1,95 @@
...
+
+	}
+
+	/**
+	 * Calculate the current Fibonacci Number with Recursion F[n-1] + F[n-2]
+	 * 
+	 * @param element
+	 * @return
+	 */
+	private int getFibonacciWithRecursion(int element) {
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_176.java]: Wrong parameter name.

File: src/main/java/am/hsp/apps/seua/task171_180/Task_179.java [hunk 34]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * 
+	 * @see main.java.am.hsp.apps.seua.base.TaskBase#solve()
+	 */
+	@Override
+	public void solve() {
+		int number_n = getUserInput("Plaese enter number N");
```
Comment (hsp-harut) [inline src/main/java/am/hsp/apps/seua/task171_180/Task_179.java]: Wrong Variable Name Format

File: src/main/java/am/hsp/apps/seua/task171_180/Task_177.java (hunk 32 above)
Comment (hsp-lusine) [reply src/main/java/am/hsp/apps/seua/task171_180/Task_177.java]: Can I change name index ???

Comment (hsp-harut) [reply src/main/java/am/hsp/apps/seua/task171_180/Task_177.java]: If you ask this question this means that you don't clearly understand the logic you implemented

**Pull request merged**